x/superfluid/keeper/stake.go

```diff
@@ -656,3 +660,209 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
+// - If valAddr provided, single delegate.
+// - If valAddr not provided and valset exists, valsetpref.Delegate
+// - If valAddr not provided and valset delegation is not possible, refer back to original lock's superfluid validator
+// - Else: error
+func (k Keeper) convertGammSharesToOsmoAndStake(
+	ctx sdk.Context,
+	sender sdk.AccAddress, valAddr string,
+	poolIdLeaving uint64, exitCoins sdk.Coins, minAmtToStake sdk.Int, originalSuperfluidValAddr string,
+) (totalAmtCoverted sdk.Int, err error) {
+	var nonOsmoCoins sdk.Coins
+	bondDenom := k.sk.BondDenom(ctx)
+
+	// from the exit coins, separate non-bond denom and bond denom.
+	for _, exitCoin := range exitCoins {
+		// if coin is not uosmo, add it to non-osmo Coins
+		if exitCoin.Denom != bondDenom {
+			nonOsmoCoins = append(nonOsmoCoins, exitCoin)
+		}
+	}
+	originalBondDenomAmt := exitCoins.AmountOf(bondDenom)
+
+	// track how much non-uosmo tokens we have converted to uosmo
+	totalAmtCoverted = sdk.ZeroInt()
+
+	// iterate over non-bond denom coins and swap them into bond denom
+	for _, coinToConvert := range nonOsmoCoins {
+		tokenOutAmt, err := k.pmk.SwapExactAmountIn(ctx, sender, poolIdLeaving, coinToConvert, bondDenom, sdk.ZeroInt())
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		totalAmtCoverted = totalAmtCoverted.Add(tokenOutAmt)
+	}
+
+	// add the converted amount with the amount of osmo from exit coin to get total amount we would be staking
+	totalAmtToStake := originalBondDenomAmt.Add(totalAmtCoverted)
+
+	// check if the total amount to stake after all conversion is greater than provided min amount to stake
+	if totalAmtToStake.LT(minAmtToStake) {
+		return sdk.ZeroInt(), types.TokenConvertedLessThenDesiredStakeError{
+			ActualTotalAmtToStake:   totalAmtToStake,
+			ExpectedTotalAmtToStake: minAmtToStake,
+		}
+	}
+
+	// var val stakingtypes.Validator
+	// if given valAddr is empty, we use delegation preference given from valset-pref module or reference from superfluid staking
+	if valAddr == "" {
+		err := k.vspk.DelegateToValidatorSet(ctx, sender.String(), sdk.NewCoin(bondDenom, totalAmtToStake))
+		// if valset-pref delegation errored due to no existing delegation existing, fall back and try using superfluid staked validator
+		if err == valsettypes.ErrNoDelegation {
+			val, err := k.validateValAddrForDelegate(ctx, originalSuperfluidValAddr)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+
+			// delegate now!
+			_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		} else if err != nil { // for other errors, handle error
+			return sdk.ZeroInt(), err
+		}
+	} else {
+		val, err := k.validateValAddrForDelegate(ctx, valAddr)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+		// delegate now!
+		_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+	}
```

Comment:
nit: this could be refactored to reduce code duplication of:

```go
		val, err := k.validateValAddrForDelegate(ctx, valAddr)
		if err != nil {
			return sdk.ZeroInt(), err
		}
		// delegate now!
		_, err = k.sk.Delegate(ctx, sender, totalAmtToStake, stakingtypes.Unbonded, val, true)
		if err != nil {
			return sdk.ZeroInt(), err
		}
```

Let me know if a more specific suggestion could help here

Reply:
addressed in https://github.com/osmosis-labs/osmosis/pull/5949/commits/96f486b5e1bbf468406fc4b5d1aa3e17ef6b4003, also added test cases